crates/ruff_server/src/server/api/requests/format.rs

```diff
@@ -39,6 +39,19 @@ pub(super) fn format_document(snapshot: &DocumentSnapshot) -> Result<super::Form
     if formatted == source {
         return Ok(None);
     }
+
+    // special case - avoid adding a newline to a notebook cell if it didn't already exist
+    if snapshot.query().as_notebook().is_some() {
+        let mut trimmed = formatted.as_str();
+        if !source.ends_with('\n') {
+            trimmed = trimmed.trim_end_matches('\n');
+        }
+        if !source.ends_with('\r') {
+            trimmed = trimmed.trim_end_matches('\r');
+        }
```

Comment:
Does this handle the `\r\n` case?

Reply:
It should, since we strip the `\n` first and then the `\r`.